src/modbus-rtu.c

```diff
@@ -272,32 +272,73 @@ static ssize_t _modbus_rtu_send(modbus_t *ctx, const uint8_t *req, int req_lengt
 #if defined(_WIN32)
     modbus_rtu_t *ctx_rtu = ctx->backend_data;
     DWORD n_bytes = 0;
-    return (WriteFile(ctx_rtu->w_ser.fd, req, req_length, &n_bytes, NULL)) ? (ssize_t)n_bytes : -1;
+    ssize_t rv;
+
+    rv = WriteFile(ctx_rtu->w_ser.fd, req, req_length, &n_bytes, NULL) ? (ssize_t)n_bytes : -1;
+
+    /* if local echo workaround is enabled, try to discard the count of bytes
+     *  we sent ourself */
+    if (ctx_rtu->suppress_echo && n_bytes > 0) {
+        /* we don't use a timeout here since we should have our own caused stuff returned
+         * immediately - and we know the internals of win32_ser_select from code above */
+        if (win32_ser_select(&ctx_rtu->w_ser, (int)n_bytes, NULL) <= 0) {
+             /* propagate errors and timeouts */
+             return -1;
+        } else {
+             /* discard the count of bytes we caused ourself, but keep possible bytes
+              * of following packet we already received */
+             unsigned int n = ctx_rtu->w_ser.n_bytes;
+
+             if (n > (unsigned int)n_bytes) {
+                 n = (unsigned int)n_bytes;
+             }
+
+             ctx_rtu->w_ser.n_bytes -= n;
+        }
+    }
+    return rv;
 #else
-#if HAVE_DECL_TIOCM_RTS
     modbus_rtu_t *ctx_rtu = ctx->backend_data;
-    if (ctx_rtu->rts != MODBUS_RTU_RTS_NONE) {
-        ssize_t size;
+    ssize_t size;
 
+#if HAVE_DECL_TIOCM_RTS
+    if (ctx_rtu->rts != MODBUS_RTU_RTS_NONE) {
         if (ctx->debug) {
             fprintf(stderr, "Sending request using RTS signal\n");
         }
 
         ctx_rtu->set_rts(ctx, ctx_rtu->rts == MODBUS_RTU_RTS_UP);
         usleep(ctx_rtu->rts_delay);
+    }
+#endif
 
-        size = write(ctx->s, req, req_length);
+    size = write(ctx->s, req, req_length);
 
+#if HAVE_DECL_TIOCM_RTS
+    if (ctx_rtu->rts != MODBUS_RTU_RTS_NONE) {
         usleep(ctx_rtu->onebyte_time * req_length + ctx_rtu->rts_delay);
         ctx_rtu->set_rts(ctx, ctx_rtu->rts != MODBUS_RTU_RTS_UP);
-
-        return size;
-    } else {
-#endif
-        return write(ctx->s, req, req_length);
-#if HAVE_DECL_TIOCM_RTS
     }
 #endif
+
+    /* if local echo workaround is enabled, try to discard the count of bytes
+     *  we sent ourself */
+    if (ctx_rtu->suppress_echo && size > 0) {
+        uint8_t echo[MODBUS_RTU_MAX_ADU_LENGTH];
+        ssize_t r_size = 0;
+
+        while (r_size < size) {
```

Comment:
right, you have a loop here for linux. (I don't know win api well enough)

Reply:
I find the variable naming of this section confusing (echo, c, r_size, size).